src/watchman_client.js

```
@@ -0,0 +1,341 @@
+   * of the 'watch' or 'watch-project' here.
+   */
+  function watch(subscription, root) {
+    var self = this;
+    
+    let promise = new Promise((resolve, reject) => {
+      let watcherInfo = getWatcherInfo(subscription);
+
+      if (_relative_root) { 
+        _client.command(['watch-project', root], (error, resp) => {
+          if (error) {
+            reject(error);
+          } else {
+            watcherInfo.root = resp.watch;
+            watcherInfo.relativePath = resp.relative_path ? resp.relative_path : '';
+            resolve(resp);
+          }
+        });
+      } else {
+        _client.command(['watch', root], (error, resp) => {
+          if (error) {
+            reject(error);
+          } else {
+            watcherInfo.root = root;
+            watcherInfo.relativePath = '';
+            resolve(resp);
+          }
+        });
+      }
+    });
+    
+    return promise;
+  }
+  
+  /**
+   * Issue the 'clock' command to get the time value for use with the 'since'
+   * option during 'subscribe'.
+   */
+  function clock(subscription) {
+    var self = this;
+
+    let promise = new Promise((resolve, reject) => {
+      let watcherInfo = getWatcherInfo(subscription);
+
+      _client.command(['clock', watcherInfo.root], (error, resp) => {
+        if (error) {
+          reject(error);
+        } else {
+          watcherInfo.since = resp.clock;
+          resolve(resp);
+        }
+      });
+    });
+
+    return promise;
+  }
+
+  /**
+   * Called from WatchmanWatcher (or this object during reconnect) to create
+   * a watcherInfo entry in our _watcherMap and issue a 'subscribe' to the
+   * watchman.Client, to be handled in this object.
+   */
+  function subscribe(watchmanWatcher, root) {
+    var self = this;
+
+    let watcherInfo = createWatcherInfo(watchmanWatcher);
+    let subscription = watcherInfo.subscription;
+    
+    let promise = new Promise((resolve, reject) => {
```

this wrapper promise is not required, simply return the promise chain which results from `watch(subscription, root)...`

---

Doh! Good catch. Fixed.

---

Actually I had second thoughts on that. The reason I created the wrapper promise was because I wanted to resolve or reject that, instead of doing 'return Promise.resolve(resp)' and 'return Promise.reject(error)'.  I did switch to that, but was wondering if you thought the earlier was better, given that.